app/controllers/incomeSources/manage/ManageIncomeSourceDetailsController.scala

```diff
@@ -55,116 +55,66 @@ class ManageIncomeSourceDetailsController @Inject()(val view: ManageIncomeSource
-          case Left(_) => Future.failed(new Error(s"Unexpected exception incomeSourceIdHash: <$incomeSourceIdHash>"))
-          case Right(incomeSourceIdHash: IncomeSourceIdHash) =>
-
-            val hashCompareResult: Either[Throwable, IncomeSourceId] = user.incomeSources.compareHashToQueryString(incomeSourceIdHash)
-
-            hashCompareResult match {
-              case Left(exception: Exception) => Future.failed(exception)
-              case Left(_) => Future.failed(new Error(s"Unexpected exception incomeSourceIdHash: <$incomeSourceIdHash>"))
-              case Right(incomeSourceId: IncomeSourceId) =>
-                sessionService.createSession(JourneyType(Manage, SelfEmployment).toString).flatMap { _ =>
-                  sessionService.setMongoKey(ManageIncomeSourceData.incomeSourceIdField, incomeSourceId.value, JourneyType(Manage, SelfEmployment)).flatMap {
-                    case Right(_) => handleRequest(
-                      sources = user.incomeSources,
-                      isAgent = false,
-                      backUrl = controllers.incomeSources.manage.routes.ManageIncomeSourceController.show(false).url,
-                      incomeSourceIdHashMaybe = Some(incomeSourceIdHash),
-                      incomeSourceType = SelfEmployment
-                    )
-                    case Left(exception) => Future.failed(exception)
-                  }
-                }.recover {
-                  case ex =>
-                    Logger("application").error(s"[ManageIncomeSourceDetailsController][showSoleTraderBusiness] - ${ex.getMessage} - ${ex.getCause}")
-                    itvcErrorHandler.showInternalServerError()
-                }
-            }
+      withSessionData(JourneyType(Manage, incomeSourceType), InitialPage) { _ =>
+        incomeSourceType match {
+          case SelfEmployment => id match {
+            case Some(realId) => showSoleTrader(realId, isAgent)
+            case None => Logger("application")
+              .error(s"[ManageIncomeSourceDetailsController][show] no incomeSourceId supplied with SelfEmployment isAgent = $isAgent")
+              Future.successful(if (isAgent) {
+                itvcErrorHandlerAgent.showInternalServerError()
+              } else {
+                itvcErrorHandler.showInternalServerError()
+              })
+          }
+          case _ => handleRequest(
+            sources = user.incomeSources,
+            isAgent = isAgent,
+            incomeSourceIdHashMaybe = None,
+            backUrl = controllers.incomeSources.manage.routes.ManageIncomeSourceController.show(isAgent).url,
+            incomeSourceType = incomeSourceType
+          )
         }
       }
   }
 
-  def showSoleTraderBusinessAgent(hashIdString: String): Action[AnyContent] = auth.authenticatedAction(isAgent = true) {
-    implicit mtdItUser =>
-      withIncomeSourcesFS {
+  def showSoleTrader(hashIdString: String, isAgent: Boolean)(implicit user: MtdItUser[_]): Future[Result] = {
```

Comment:
⛏️ Maybe rename to something like handleSoleTrader? Having this as showSoleTrader makes it sound like a route